### PR TITLE
Mech strafing (Atomized edition)

### PR DESCRIPTION
**Pull request by MLGTASTICa**
## What Does This PR Do
Adds mech strafing , mechs can strafe sideways or backwards depending on the flags given , currently , most mechs can do both , durand can only go backwards/fowards and mining ripley sideways/fowards.
Going sideways has a 60% malus , Going backwards has a 100% malus.
## Why It's Good For The Game
Makes mech gameplay far more energetic and engaging by allowing 2D movement planes.
Lets mech drivers trade off movement speed for facing the threat.

## Testing
Ran a local server , spawned mechs & strafed till i had enough 
### Declaration
- [X] I confirm that I either do not require [pre-approval](../CODE_OF_CONDUCT.md#types-of-changes-that-need-approval) for this PR, or I have obtained such approval and have included a screenshot to demonstrate this below.
https://private-user-images.githubusercontent.com/61350382/358695070-e68b69d3-03df-4300-9ebb-411c1b76a07f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQwODY0NjQsIm5iZiI6MTcyNDA4NjE2NCwicGF0aCI6Ii82MTM1MDM4Mi8zNTg2OTUwNzAtZTY4YjY5ZDMtMDNkZi00MzAwLTllYmItNDExYzFiNzZhMDdmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODE5VDE2NDkyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkxMmM0ZjA3OGQ5NDVjOWYyMjU1ZWVlNWE2NTJkN2M0ZjJjMWZiMjdiOGIzMzIyN2I3OTU0ZWU0Nzk5MzI0YzEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.oPPeGiRNBmHGUsyXUbiwrKoaRvZyRp5OH1Mo-f1CBjM
<hr>

## Changelog
:cl:
add : Mech strafing ability , mechs can now strafe backwards and sideways , the only exceptions are the Durand,  which can only strafe backwards , and the mining ripley , which can only strafe sideways.
/:cl:

### CODE REVIEW

File: code/game/mecha/mecha.dm [hunk 1]
```diff
@@ -44,6 +44,11 @@
 	var/emp_proof = FALSE //If it is immune to emps
 	var/emag_proof = FALSE //If it is immune to emagging. Used by CC mechs.
 
+	/// Wheter the mech moves in a strafing motion
```
Comment (Sirryan2002) [inline code/game/mecha/mecha.dm:47]:
```suggestion
	/// Whether the mech moves in a strafing motion
```

File: code/game/mecha/mecha.dm [hunk 2]
```diff
@@ -44,6 +44,11 @@
...
+	var/strafing = FALSE
+	/// The available modes of strafing
```
Comment (Sirryan2002) [inline code/game/mecha/mecha.dm:49]:
```suggestion
	/// Controls how the Mech is able to Strafe.
```

File: code/game/mecha/mecha.dm [hunk 3]
```diff
@@ -326,19 +332,33 @@
 		step_in = initial(step_in)
 	var/move_result = 0
 	var/move_type = 0
+	var/step_malus = 0
```
Comment (Sirryan2002) [inline code/game/mecha/mecha.dm:335]: maybe add a comment detailing what this local var is for? Not sure what it's purpose is just looking at the name

File: code/game/mecha/mecha.dm [hunk 4]
```diff
@@ -326,19 +332,33 @@
...
 	if(internal_damage & MECHA_INT_CONTROL_LOST)
 		move_result = mechsteprand()
 		move_type = MECHAMOVE_RAND
 	else if(dir != direction)
-		move_result = mechturn(direction)
-		move_type = MECHAMOVE_TURN
+		if(strafing)
+			var/old_dir = dir
+			if(strafing_flags & MECH_STRAFING_BACKWARDS && direction == REVERSE_DIR(dir))
```
Comment (Sirryan2002) [inline code/game/mecha/mecha.dm:342]:
```suggestion
			if((strafing_flags & MECH_STRAFING_BACKWARDS) && direction == REVERSE_DIR(dir))
```
For Readability Sake

File: code/game/mecha/mecha.dm [hunk 5]
```diff
@@ -326,19 +332,33 @@
...
+				move_result = mechstep(direction)
+				move_type = MECHAMOVE_STEP
+				dir = old_dir
+				step_malus = step_in
+			else if(strafing_flags & MECH_STRAFING_SIDEWAYS && direction & (turn(dir,90) | turn(dir,-90)))
```
Comment (Sirryan2002) [inline code/game/mecha/mecha.dm:347]:
```suggestion
			else if((strafing_flags & MECH_STRAFING_SIDEWAYS) && (direction & (turn(dir, 90) | turn(dir, -90))))
```
readbility

File: code/game/mecha/mecha_actions.dm [hunk 6]
```diff
@@ -241,3 +244,30 @@
...
+
+/datum/action/innate/mecha/mech_strafing_mode/Grant(mob/living/L, obj/mecha/M)
+	. = ..()
+	button_overlay_icon_state = "[initial(button_overlay_icon_state)]_[M.strafing ? "on" : "off"]"
+	UpdateButtons()
+
+/datum/action/innate/mecha/mech_strafing_mode/Activate(force = FALSE)
+	if(force && chassis)
+		chassis.strafing = !chassis.strafing
+		button_overlay_icon_state = "[initial(button_overlay_icon_state)]_[chassis.strafing ? "on" : "off"]"
```
Comment (lewcc) [inline code/game/mecha/mecha_actions.dm:261]: I don't have the code on me rn, but I think you can include this in `UpdateButton()` for the action, or whatever proc gets called within `UpdateButtons()` so you don't manually need to set the icon state each time,

File: code/game/mecha/mecha_actions.dm [hunk 7]
```diff
@@ -241,3 +244,30 @@
...
+		UpdateButtons()
+		return
+	if(!owner || !chassis || chassis.occupant != owner)
+		return
+	if(chassis.strafing)
+		chassis.occupant_message("You toggle off \the [chassis]'s strafing mode")
+		chassis.strafing = FALSE
+	else
+		chassis.occupant_message("You toggle on \the [chassis]'s strafing mode")
```
Comment (lewcc) [inline code/game/mecha/mecha_actions.dm:270]:
```suggestion
		chassis.occupant_message("You disable [chassis]'s strafing mode.")
		chassis.strafing = FALSE
	else
		chassis.occupant_message("You enable [chassis]'s strafing mode.")
```

**Pull request closed without merging**